cmd/cc-uploader/main.go

```diff
@@ -189,16 +131,87 @@ func initializeServer(logger lager.Logger, uploaderConfig config.UploaderConfig,
 			tls.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
 		}
 
-		server = &http.Server{
-			Addr:      uploaderConfig.MutualTLS.ListenAddress,
-			Handler:   ccUploaderHandler,
-			TLSConfig: clientTLSConfig,
-		}
-		return server, http_server.NewTLSServer(uploaderConfig.MutualTLS.ListenAddress, ccUploaderHandler, clientTLSConfig)
+		return http_server.NewTLSServer(uploaderConfig.MutualTLS.ListenAddress, ccUploaderHandler, clientTLSConfig)
 	}
-	server = &http.Server{
-		Addr:    uploaderConfig.ListenAddress,
-		Handler: ccUploaderHandler,
+
+	return http_server.New(uploaderConfig.ListenAddress, ccUploaderHandler)
+}
+
+func waitForDrainingToFinish() <-chan struct{} {
+	done := make(chan struct{})
+	go func() {
+		uploadWaitGroup.Wait() // wait for all in-flight uploads to call Done()
+		close(done)            // signal completion
+	}()
+	return done
+}
+
+func configureServers(logger lager.Logger, uploaderConfig config.UploaderConfig, reconfigurableSink *lager.ReconfigurableSink) ifrit.Process {
+
+	var nonTLSRunner ifrit.Runner
+	tlsRunner := initializeServer(logger, uploaderConfig, true)
+	members := grouper.Members{
+		{Name: "cc-uploader-tls", Runner: tlsRunner},
+	}
+	if !uploaderConfig.DisableNonTLS {
+		nonTLSRunner = initializeServer(logger, uploaderConfig, false)
+		members = append(grouper.Members{
+			{Name: "cc-uploader", Runner: nonTLSRunner},
+		}, members...)
+	}
+	if uploaderConfig.DebugServerConfig.DebugAddress != "" {
+		members = append(grouper.Members{
+			{Name: "debug-server", Runner: debugserver.Runner(uploaderConfig.DebugServerConfig.DebugAddress, reconfigurableSink)},
+		}, members...)
+	}
+
+	group := grouper.NewOrdered(os.Interrupt, members)
+	monitor := ifrit.Invoke(sigmon.New(group))
+	logger.Info("ready")
+
+	return monitor
+}
+
+func main() {
+	runtime.GOMAXPROCS(runtime.NumCPU())
```

Comment:
Unless there is a strong reason to set this, let the Go runtime decide.

From the docs:
> Otherwise, the Go runtime selects an appropriate default value from a combination of
>  * the number of logical CPUs on the machine,
>  * the process’s CPU affinity mask,
>  * and, on Linux, the process’s average CPU throughput limit based on cgroup CPU quota, if any.

Especially the last one (cgroups) can otherwise be tricky.

Reply:
True, seems to be an old relict..